### ISSUE TITLE
[BUG] arrayfire::locate fails for large arrays

### ISSUE THREAD

**Issue by BA8F0D39** [arrayfire/arrayfire#3372]
Description
===========
arrayfire::locate fails for input large arrays. I only tested this on the rust version of arrayfire. No sure if it appears in the C++ version.

Reproducible Code and/or Steps
------------------------------

For a large array of A, arrayfire locate fails
```

arrayfire::set_seed(100);
let A_dims = arrayfire::Dim4::new(&[107500,1,32000,1]);
let mut A = arrayfire::randu::<bool>(A_dims);

let result = arrayfire::locate(&A);

println!("  result.dims()[0] {}",result.dims()[0]);
println!("  result.dims()[1] {}",result.dims()[1]);
println!("  result.dims()[2] {}",result.dims()[2]);	
```
```
[1]    84261 segmentation fault (core dumped) 
```
System Information
------------------
ArrayFire v3.9.0 (CUDA, 64-bit Linux, build 715e21fcd)
Platform: CUDA Runtime 12.0, Driver: 525.89.02
[0] NVIDIA GeForce RTX 3090, 24260 MB, CUDA Compute 8.6
Info String:
ArrayFire v3.9.0 (CUDA, 64-bit Linux, build 715e21fcd)
Platform: CUDA Runtime 12.0, Driver: 525.89.02
[0] NVIDIA GeForce RTX 3090, 24260 MB, CUDA Compute 8.6
Arrayfire version: (3, 9, 0)
Name: NVIDIA_GeForce_RTX_3090
Platform: CUDA
Toolkit: v12.0
Compute: 8.6
Revision: 715e21fcd

Checklist
---------
- [x] Using the latest available ArrayFire release
- [x] GPU drivers are up to date

**Comment by 9prady9** [comment]
@syurkevi Have we been able to reproduce this locally ? 
 
 @BA8F0D39 If you are sure it is core library issue, there is no need for having separate issue open for the same problem in bindings repository. https://github.com/arrayfire/arrayfire-rust/issues/373 Please clarify, thank you

**Comment by BA8F0D39** [comment]
@9prady9 
Ok, I'll close this

**Comment by syurkevi** [comment]
I've found a few kernels this week that fail for larger sizes, I think this may be related to those failures as well. I will re-open the issue if I'm able to reproduce in the core lib.

**Comment by 9prady9** [comment]
@BA8F0D39 My comment was misunderstood, I didn't mean to close the core library issue rather close the [bindings](https://github.com/arrayfire/arrayfire-rust/issues/373) one , which I did just now. I am reopening this issue since you reported it again in the bindings repo.

**Comment by BA8F0D39** [comment]
@syurkevi 
Did you manage to replicate it?

**Comment by syurkevi** [comment]
Yes, this issue is in the core repo in one of the where kernels [here](https://github.com/syurkevi/arrayfire/blob/e7ffdd080eae16b2d517a5e41feab5c39a748bc1/src/backend/cuda/kernel/where.hpp#LL63C31-L63C34) as well as in the scan kernels.
This is a larger issue due to the use of uint instead of a 64bit type in many of our cuda kernels internally. This choice was originally made for performance reasons but needs to be revisited.
 Some initial fixes can be found on this [branch](https://github.com/arrayfire/arrayfire/compare/master...syurkevi:arrayfire:bigarrays) but a reworking of additional kernels such as where is still outstanding. And the JIT will need some careful reworking as well. 
Once fixed this should also enable https://github.com/arrayfire/arrayfire-rust/issues/284  . There is some other issue I can't find right now specifically about reduce with large arrays that needs to be referenced here as well (perhaps it was deleted?)

**Comment by BA8F0D39** [comment]
@syurkevi 
Some functions already take in the index as input

af_where (af_array *idx, const af_array in)

Can't you detect if idx is 32bit or 64bit and use the appropriate kernel?

If you pass idx that is 32bit, arrayfire will generate a kernel that uses 32bit indexing.

If you pass idx that is 64bit, arrayfire will generate a kernel that uses 64bit indexing.

Alternatively, you can use other integer types for indexing.
